vividus-plugin-mobitru/src/main/java/org/vividus/mobitru/client/MobitruFacadeImpl.java

```diff
@@ -54,39 +66,101 @@ public MobitruFacadeImpl(MobitruClient mobitruClient)
+            return takeDevice(foundDevices);
+        }
+        Device device = new Device();
+        device.setDesiredCapabilities(desiredCapabilities.asMap());
+        Waiter deviceWaiter = new DurationBasedWaiter(new WaitMode(waitForDeviceTimeout, RETRY_TIMES));
+        return takeDevice(device, deviceWaiter);
+    }
+
+    @Override
+    public void installApp(String deviceId, String appRealName) throws MobitruOperationException
+    {
+        mobitruClient.installApp(deviceId, findApp(appRealName));
+    }
+
+    @Override
+    public void returnDevice(String deviceId) throws MobitruOperationException
+    {
+        mobitruClient.returnDevice(deviceId);
+    }
+
+    private String takeDevice(Device device, Waiter deviceWaiter) throws MobitruOperationException
+    {
+        LOGGER.info("Trying to take device with configuration {}", device);
+        String capabilities = performMapperOperation(mapper -> mapper.writeValueAsString(device));
+        byte[] receivedDevice = deviceWaiter.wait(() -> {
             try
             {
                 return mobitruClient.takeDevice(capabilities);
             }
-            catch (MobitruDeviceSearchException e)
+            catch (MobitruDeviceTakeException e)
             {
                 LOGGER.warn("Unable to take device, retrying attempt.", e);
                 return null;
             }
         }, Objects::nonNull);
         if (null == receivedDevice)
         {
-            throw new MobitruOperationException(String.format("Unable to find device using %s", desiredCapabilities));
+            throw new MobitruDeviceTakeException(String.format("Unable to take device with configuration %s", device));
         }
-        Device device = performMapperOperation(mapper -> mapper.readValue(receivedDevice, Device.class));
-        return (String) device.getDesiredCapabilities().get("udid");
+        Device takenDevice = performMapperOperation(mapper -> mapper.readValue(receivedDevice, Device.class));
+        LOGGER.info("Device with configuration {} is taken", takenDevice);
+        return (String) takenDevice.getDesiredCapabilities().get(UDID);
     }
 
-    @Override
-    public void installApp(String deviceId, String appRealName) throws MobitruOperationException
+    private String takeDevice(List<Device> devices) throws MobitruOperationException
     {
-        mobitruClient.installApp(deviceId, findApp(appRealName));
+        Waiter waiter = new RetryTimesBasedWaiter(Duration.ZERO, 1);
+        Iterator<Device> devicesIterator = devices.iterator();
+        while (true)
+        {
+            try
+            {
+                return takeDevice(devicesIterator.next(), waiter);
+            }
+            catch (MobitruDeviceTakeException e)
+            {
+                if (!devicesIterator.hasNext())
+                {
+                    throw e;
+                }
+                LOGGER.atWarn().log(e.getMessage());
```

Comment:
why warn and not error?

Reply:
because we still have a chance to try another device